Added a Molokai ColorScheme

challsted opened the pull request:
Added a new [Molokai](https://github.com/tomasr/molokai) based ColorScheme with 2 settings:

* Default
   * Normal Mode is Cyan
   * Insert Mode is Green
   * Visual Mode is Yellow
   * Replace Mode is Red
* Alternate
   * Normal Mode is Pink
   * Insert Mode is Green
   * Visual Mode is Orange
   * Replace Mode is Red

Also updated documentation to reflect:

1. This new colorscheme
2. Details for future contributors of colorschemes that are currently undocumented
  * This was needed due to the appearance of feature creep

itchyny commented:
- I prefer lower case for the color hex.
- Please specify cterm colors if possible.

challsted commented:
@itchyny I've converted to lowercase hex values. But I'm not understanding what you mean by "cterm colors"
As mentioned, I originally copied from the Dracula Color Scheme and just changed the values from there. 
I had attempted to do so with the "landscape" color scheme as was suggested in the help doc but I was unable to figure out how the colors were being set. 

If you don't mind, pointing me in the direction of how to set cterm colors, I will be happy to make the needed changes.

challsted commented:
So, reading through the help doc again, i see what i think is a provided way to make both gui and cui (cui == cterm?) colors. Ill work with that soon and test it out. 

I guess it would be important to mention ive only tested this in NeoVim CLI inside TMux. 

Probably best to make sure it works in Vim 7.4, 8, and both inside and outside TMux, as well as in GVim. I dont have OSX so i cant test it in MacVim.

itchyny commented:
Sorry for my confusing comment. I committed to the Dracula colorscheme recently so that it uses the terminal colours properly. The  `lightline#colorscheme#fill` utility automatically fills the terminal colours converted from the hex colours, which might be a great feature, but it means that the lightline colorscheme ignores the colors of the original colorscheme. If the original colorscheme defined the ctermfg/ctermbg, it's preferable to set the colours in the lightline colorscheme too. If it's not clear, no worry, I'll merge this pull request and take care of them later.

challsted commented:
Ok, i see what your asking now, but to be honest, i dont understand what is needed to be done. The numbers added seem to be a value between 0 and 255. Although testing showed that any value can be inserted with seemingly random effect (quick tests with numbers like 256, 300, 999, 100000).

But to answer your question, yes the Molokai theme i based these colors off of does have ctermfg and ctermbg settings. I can do another round of testing and add these. Its all a learning process for me and im really enjoying it so far! Thanks fol taking time to reply to make sure this PR is good :+1:

challsted commented:
@itchyny ok, i THINK i've got everything now :)
if you could please have a look, im having issues testing for some reason, i can see the origional colors just fine, but even when switching colors that should be obvious (orange to pink for example) i was not able to see the changes even when closing and relaunching nvim. 

Let me know what you think

itchyny commented:
Looks good, I'll test on my environment and merge this pull request soon.

For reference https://github.com/guns/xterm-color-table.vim is a good plugin.

challsted commented:
That plugin would of been helpful! :100: I was kind of poking into the dark until there at the end where i googled it...

itchyny commented on autoload/lightline/colorscheme/molokai.vim:
```diff
@@ -0,0 +1,34 @@
+let s:p.insert.left = [ [ s:black, s:green ], [ s:cyan, s:black ] ]
+let s:p.visual.left = [ [ s:black, s:orange ], [ s:black, s:cyan ] ]
+let s:p.replace.left = [ [ s:black, s:red ], [ s:black, s:orange ] ]
+let s:p.normal.left = [ [ s:black, s:pink ], [ s:orange, s:black ] ]
```
I would like to make the appearance of colorschemes consistent (at least for newly created colorschemes). Would you change the colours like https://github.com/itchyny/lightline.vim#wombat-with-the-patched-font, which means cyan for normal mode, green for insert mode, yellow for visual mode and red for replace mode.

challsted replied:
Would you be opposed to some kind of setting to allow standard and more Molokai like? (Not sure how to word this) in the color scheme would be a if(flag)
Current colors
Else
Standard colors
Endif

Or however color schemes handle this type of logic (i have examples inside of the currunt molokai color scheme. 

This would allow for your standard, but if a user wanted to use more "molokai-like" colors they could. 

If your cool with this, i can do this in here, or in a 2nd pull request

challsted commented:
@itchyny please review again and let me know if anything else needs to be updated

itchyny commented:
I'm not sure what you mean by `molokai-like`. Molokai is a colorscheme and has nothing to do with the colours of the statusline in each modes, right?

challsted commented:
Sorry, my wording was off. 

What i mean is that Molokai has 4 primary colors:
* Pink
* Green
* Blue
* Orange

When i say "molokai-like" i mean use those colors as often as possible.

challsted commented:
@itchyny i understand your busy, but have you had anytime to review this again? its been 2 days since i asked last.

itchyny commented:
I'm sorry. I don't want to add a configuration flag in the colorschemes and I was wondering how should I deal with this pull request...

challsted commented:
How is it different then say a light or dark? 
Its simply defaulting to the way you want it, and if a user manually updates a value in there Vimrc then they can get a slightly different (alternate) way. 

I understand your having issues, would just like to communicate about it :)

itchyny commented:
The `background` options is known to many people using Vim and easy to explain its behavior but the `g:lightline_molokai_alternative` option is hard to know what it is and also hard for me to add ducument in the lightline help. Also I've keep this plugin configurable by single global variable and hesitate to introduce a new option like this.

challsted commented:
I see your point. 
I can remove it if that would make this PR end TBH, was just trying to keep my original idea and more strict to what Molokai is.

challsted commented:
@itchyny ok, removed flag, and i switched some colors around, i like it like this better now

had it fail on me a few times when loading, found some cool features in powerline (insert.middle specifically) and played with using .middles to help notify the user of mode better. Also worked with adding in more pink. 

Please review it as soon as you have time.

challsted commented:
@itchyny Thank you! :+1: :100:
